File: ext/gsl/tamu_anova.c
#include "include/rb_gsl.h"

#ifdef HAVE_TAMU_ANOVA_TAMU_ANOVA_H
VALUE rb_tamu_anova_alloc(int argc, VALUE *argv, VALUE klass)
{
  gsl_vector *data;
  gsl_vector_long *factor;
  long I, J;
  struct tamu_anova_table *table;
  switch (argc) {
  case 3:
  case 4:
    Data_Get_Struct(argv[0], gsl_vector, data);
    Data_Get_Struct(argv[1], gsl_vector_long, factor);
    if (argc == 3) {
      I = data->size;
      J = NUM2INT(argv[2]);
    } else {
      I = NUM2INT(argv[2]);
      J = NUM2INT(argv[3]);
    }
    table = (struct tamu_anova_table *) malloc(sizeof(struct tamu_anova_table));
    *table = tamu_anova(data->data, factor->data, I, J);
    break;
  default:
    rb_raise(rb_eArgError, "Wrong number of arguments (%d for 3 or 4)", argc);
    break;
  }
  return Data_Wrap_Struct(klass, 0, free, table);
}

VALUE rb_tamu_anova_printtable(VALUE *vTable)
{
  struct tamu_anova_table *table;
  Data_Get_Struct(vTable, struct tamu_anova_table, table);
  tamu_anova_printtable(*table);
  return Qtrue;
}

#endif

void Init_tamu_anova(VALUE module)
{
#ifdef HAVE_TAMU_ANOVA_TAMU_ANOVA_H
  VALUE mTAMU_ANOVA;
  VALUE cTable;

  mTAMU_ANOVA = rb_define_module_under(module, "TAMU_ANOVA");
  cTable = rb_define_class_under(mTAMU_ANOVA, "Table", cGSL_Object);

  rb_define_singleton_method(cTable, "alloc", rb_tamu_anova_alloc, -1);
  rb_define_singleton_method(cTable, "oneway", rb_tamu_anova_alloc, -1);

  rb_define_method(cTable, "print", rb_tamu_anova_printtable, 0);
#endif
}

File: examples/tamu_anova.rb
#!/usr/bin/env ruby
require("gsl")

data = GSL::Vector[88.60,73.20,91.40,68.00,75.20,63.00,53.90,69.20,
50.10,71.50,44.90,59.50,40.20,56.30,38.70,31.00,
39.60,45.30,25.20,22.70]
factor = GSL::Vector::Int[1,1,1,1,1,2,2,2,2,2,3,3,3,3,3,4,4,4,4,4]

table = GSL::TAMU_ANOVA::Table.oneway(data, factor, 4)
table.print

data = GSL::Vector[45.50,45.30,45.40,44.40,44.60,43.90,44.60,44.00,44.20,
43.90,44.70,44.20,44.00,43.80,44.60,43.10,46.00,45.90,
44.80,46.20,45.10,45.50]
factor = GSL::Vector::Int[1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,3,3,3,3,3,3]

table = GSL::TAMU_ANOVA::Table.oneway(data, factor, 3)
table.print
